fix(techdocs): add missing i18n support for techdocs plugin

its-mitesh-kumar opened the pull request:
## Description 
Added i18n support for missing strings for TechDocs Plugin.

## UI after changes

https://github.com/user-attachments/assets/1548fdd9-ea5e-4d84-b067-76ac745e5ac9

# Steps to Test TechDocs Locally

Follow the official [Backstage TechDocs Getting Started](https://backstage.io/docs/features/techdocs/getting-started) documentation.

## 1. Add Configuration

Add the following configuration in `app-config.yaml`:

```yaml
techdocs:
  builder: 'local'
  publisher:
    type: 'local'
  generator:
    runIn: local
```

## 2. Create Python Virtual Environment

```bash
python3 -m venv techdocs-venv
```

## 3. Activate Virtual Environment

```bash
source techdocs-venv/bin/activate
```

## 4. Install MkDocs (One-time only)

```bash
pip install mkdocs mkdocs-techdocs-core
```

## 5. Install Additional Dependencies

```bash
brew install graphviz plantuml
```

## 6. Start Backstage

**Note**: Run this in the same terminal where the virtual environment is activated.

```bash
yarn start
```



<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

its-mitesh-kumar committed:
fix(techdocs): add missing i18n support for techdocs plugin

Signed-off-by: its-mitesh-kumar <itsmiteshkumar98@gmail.com>

its-mitesh-kumar committed:
adding changset

Signed-off-by: its-mitesh-kumar <itsmiteshkumar98@gmail.com>

christoph-jerolimov reviewed (approved):
Hi @its-mitesh-kumar thanks for this work :+1:

Code looks good to me, can you please rebase this to trigger the builds again?

You will be also required that you rebuild the `report-alpha.api.md` file with `yarn tsc && yarn build:all && yarn build:api-reports:only`.

its-mitesh-kumar committed:
updating the api report

Signed-off-by: its-mitesh-kumar <itsmiteshkumar98@gmail.com>

its-mitesh-kumar committed:
updating the test file

Signed-off-by: its-mitesh-kumar <itsmiteshkumar98@gmail.com>

christoph-jerolimov reviewed (changes requested):
Hi @its-mitesh-kumar thanks for your PR. The result looks good but I have an ask about the exported types. Can you please take a look? Thanks :)

its-mitesh-kumar committed:
addressing the comments

Signed-off-by: its-mitesh-kumar <itsmiteshkumar98@gmail.com>

its-mitesh-kumar committed:
updating api-report

Signed-off-by: its-mitesh-kumar <itsmiteshkumar98@gmail.com>

its-mitesh-kumar commented:
@christoph-jerolimov @alexlorenzi @byan1197 @PeaWarrior @johnphilip283 @squid-ney @Eswaraiahsapram Please do review the PR

its-mitesh-kumar commented:
@christoph-jerolimov @alexlorenzi @byan1197 @PeaWarrior @johnphilip283 @squid-ney @Eswaraiahsapram @benjdlambert  Please review and merge the PR.

schultzp2020 commented on .changeset/vast-terms-argue.md:
```diff
@@ -0,0 +1,5 @@
+'@backstage/plugin-techdocs': patch
+---
+
+Added missing i18n
```
Can you slightly expand on the changes include in this PR?

schultzp2020 reviewed (commented):
One minor comment; otherwise LGTM! 🛶

its-mitesh-kumar committed:
updating changeset

Signed-off-by: its-mitesh-kumar <itsmiteshkumar98@gmail.com>

schultzp2020 commented on plugins/techdocs/src/home/components/Tables/types.ts:
```diff
@@ -29,3 +31,40 @@ export type DocsTableRow = {
+  createStarEntityAction(
+    isStarredEntity: Function,
+    toggleStarredEntity: Function,
+    t: Function,
```
```suggestion
t?: Function,
```

Optional here as well, matching the suggestion above.

schultzp2020 commented on plugins/techdocs/src/home/components/Tables/actions.tsx:
```diff
@@ -17,18 +17,27 @@
       return {
         icon: () => <ShareIcon fontSize="small" />,
-        tooltip: 'Click to copy documentation link to clipboard',
+        tooltip: t('table.actions.copyDocsUrl', {}),
```
```suggestion
  createCopyDocsUrlAction(
    copyToClipboard: Function,
    t?: InternalTranslationFunctionType,
  ) {
    return (row: DocsTableRow) => {
      return {
        icon: () => <ShareIcon fontSize="small" />,
        tooltip: t?.('table.actions.copyDocsUrl', {}) ?? 'Click to copy documentation link to clipboard',
```

With `t` optional, the factory gracefully falls back to the original English string when called by consumers who haven't adopted i18n yet.

schultzp2020 reviewed (commented):
Thanks for the comprehensive work adding i18n support across the TechDocs plugin, @its-mitesh-kumar!

### Backwards compatibility of action factory signatures

The main thing I'd like to see addressed before merging is that the `createCopyDocsUrlAction` and `createStarEntityAction` methods now **require** a `t` translation function parameter. Previously their signatures were:

```ts
createCopyDocsUrlAction(copyToClipboard: Function)
createStarEntityAction(isStarredEntity: Function, toggleStarredEntity: Function)
```

Since these are reachable through the `@public` exports `DocsTable.actions` and `EntityListDocsTable.actions`, any consumer calling them today without `t` will get a runtime error (`t` is `undefined`, then `t(...)` throws).

Everything else looks great!

schultzp2020 commented on plugins/techdocs/src/home/components/Tables/actions.tsx:
```diff
@@ -37,14 +46,18 @@ export const actionFactories = {
       const isStarred = isStarredEntity(entity);
+      const tooltip = isStarred
+        ? t('table.actions.removeFromFavorites', {})
+        : t('table.actions.addToFavorites', {});
```
```suggestion
  createStarEntityAction(
    isStarredEntity: Function,
    toggleStarredEntity: Function,
    t?: InternalTranslationFunctionType,
  ) {
    return (row: DocsTableRow) => {
      const entity = row.entity;
      const isStarred = isStarredEntity(entity);
      const tooltip = isStarred
        ? (t?.('table.actions.removeFromFavorites', {}) ?? 'Remove from favorites')
        : (t?.('table.actions.addToFavorites', {}) ?? 'Add to favorites');
```

Same pattern — optional `t` with English fallback defaults.

its-mitesh-kumar replied:
Addressed it.

its-mitesh-kumar committed:
marking t as optional

Signed-off-by: its-mitesh-kumar <itsmiteshkumar98@gmail.com>

its-mitesh-kumar committed:
Merge remote-tracking branch 'upstream/master' into feat/techdocs-adding-missing-i18n-support

Signed-off-by: its-mitesh-kumar <itsmiteshkumar98@gmail.com>

its-mitesh-kumar committed:
fixing prettier

Signed-off-by: its-mitesh-kumar <itsmiteshkumar98@gmail.com>

schultzp2020 commented on plugins/techdocs/src/home/components/Tables/types.ts:
```diff
@@ -15,6 +15,8 @@
 
 import { Entity, CompoundEntityRef } from '@backstage/catalog-model';
+import { TableColumn } from '@backstage/core-components';
+import { JSX } from 'react';
```
Can you fix this?

its-mitesh-kumar replied:
Done

schultzp2020 reviewed (approved):
One minor nit about explicitly exporting a type; otherwise, this PR lgtm!

its-mitesh-kumar commented:
@freben Can we get this in ?

its-mitesh-kumar committed:
use import type for type-only

Signed-off-by: its-mitesh-kumar <itsmiteshkumar98@gmail.com>

its-mitesh-kumar commented:
> One minor nit about explicitly exporting a type; otherwise, this PR lgtm!

@schultzp2020 Fixed that.

Rugvip commented on plugins/techdocs/report.api.md:
```diff
@@ -94,37 +135,8 @@ export type DocsGroupConfig = {
-  };
-};
+// @public (undocumented)
+export const DocsTable: DocsTableType;
```
Regarding the discussion in the framework SIG, this was one of the main bits. We want to avoid the separate `DocsTableType` export here, and just have it be `DocsTable`.

The new `ActionFactories` and `ColumnFactories` refactor is alright, but they need to be namespaces a bit. Probably the best fit is `DocsTableActionFactories` and `DocsTableColumnFactories`. I think that fits well with both `DocsTable` and `EntityListDocsTable`.

its-mitesh-kumar replied:
Thanks for the feedback @Rugvip! I've addressed both comments:

- Removed the separate `DocsTableType` and `EntityListDocsTableType` exports — `DocsTable` and `EntityListDocsTable` now use inline type assertions directly.
- Renamed `ActionFactories` → `DocsTableActionFactories` and `ColumnFactories` → `DocsTableColumnFactories` for better namespacing.

API reports have been updated to reflect these changes.

Rugvip commented on plugins/techdocs/report.api.md:
```diff
@@ -163,37 +185,8 @@ export type EntityListDocsGridPageProps = {
-  };
-};
+// @public (undocumented)
+export const EntityListDocsTable: EntityListDocsTableType;
```
Same thing here

Rugvip reviewed (changes requested):
Thanks! Just a few things regarding the public API surface that we discussed in the framework SIG

its-mitesh-kumar committed:
addressing the comment

Signed-off-by: its-mitesh-kumar <itsmiteshkumar98@gmail.com>

its-mitesh-kumar commented:
Thanks for the feedback @Rugvip! I've addressed both comments:

- Removed the separate `DocsTableType` and `EntityListDocsTableType` exports — `DocsTable` and `EntityListDocsTable` now use inline type assertions directly.
- Renamed `ActionFactories` → `DocsTableActionFactories` and `ColumnFactories` → `DocsTableColumnFactories` for better namespacing.

API reports have been updated to reflect these changes.

its-mitesh-kumar committed:
updating api report

Signed-off-by: its-mitesh-kumar <itsmiteshkumar98@gmail.com>

Rugvip reviewed (approved):
Thank you! 🎉, let's :shipit:
